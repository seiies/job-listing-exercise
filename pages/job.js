import React from 'react'
import { bindActionCreators } from 'redux'
import Head from 'next/head'
import Error from 'next/error'
import { initStore } from '../store'
import { fetchJobItem } from '../store/actions'
import withRoot from '../utils/withRoot'
import withRedux from '../utils/withRedux'
import Layout from '../components/Layout'
import JobDescription from '../components/JobDescription'
import PropTypes from 'prop-types'

export class JobDescriptionPage extends React.Component {
  static async getInitialProps ({ isServer, store, req, query: { id } }) {
    const jobDataInStore = store.getState().jobDescriptions[id]
    if (!jobDataInStore) {
      await store.dispatch(fetchJobItem(id))
    }

    return { id, isServer }
  }

  render () {
    const { id, jobDescriptions } = this.props
    const jobItem = jobDescriptions[id]

    if (!jobItem) {
      return <Error statusCode={404} />
    }

    return (
      <React.Fragment>
        <Head>
          <title>Job: {jobItem.title}</title>
        </Head>
        <Layout>
          <JobDescription job={jobItem} />
        </Layout>
      </React.Fragment>
    )
  }
}

JobDescriptionPage.propTypes = {
  id: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string
  ]),
  jobDescriptions: PropTypes.object.isRequired
}

const mapDispatchToProps = (dispatch) => {
  return {
    fetchJobItem: bindActionCreators(fetchJobItem, dispatch)
  }
}

export default withRedux(initStore, state => state, mapDispatchToProps)(withRoot(JobDescriptionPage))
