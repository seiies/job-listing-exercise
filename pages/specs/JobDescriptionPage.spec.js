/* global jest, describe, it, expect */

import React from 'react'
import { JobDescriptionPage } from '../job'
import { shallow } from 'enzyme'

describe('job page', () => {
  let state = { jobDescriptions: { 1: 'something' } }
  const store = {
    getState: jest.fn(() => state),
    dispatch: jest.fn()
  }
  let jobs = {
    1: {
      title: 'Job title',
      html: 'something'
    }
  }

  it('renders correctly for the existing job description', () => {
    const component = shallow(<JobDescriptionPage id={1} jobDescriptions={jobs} />)
    expect(component).toMatchSnapshot()
  })

  it('renders correctly for the wrong id', () => {
    const component = shallow(<JobDescriptionPage id={19} jobDescriptions={jobs} />)
    expect(component).toMatchSnapshot()
  })

  it('does not fetch job list in case of it is present in store', async () => {
    await JobDescriptionPage.getInitialProps({ store, query: { id: 1 } })
    expect(store.dispatch).not.toBeCalled()
  })

  it('fetches job list in case of lack of it in store', async () => {
    await JobDescriptionPage.getInitialProps({ store, query: { id: 19 } })
    expect(store.dispatch).toHaveBeenCalledTimes(1)
  })
})
