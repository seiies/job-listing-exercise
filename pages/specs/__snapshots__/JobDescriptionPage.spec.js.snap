// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`job page renders correctly for the existing job description 1`] = `
<React.Fragment>
  <SideEffect(Head)>
    <title>
      Job: 
      Job title
    </title>
  </SideEffect(Head)>
  <Component>
    <JobDescription
      job={
        Object {
          "html": "something",
          "title": "Job title",
        }
      }
    />
  </Component>
</React.Fragment>
`;

exports[`job page renders correctly for the wrong id 1`] = `
<Error
  statusCode={404}
/>
`;
